### DISCUSSION TITLE
Upload component

### DISCUSSION THREAD

**Discussion by iveneva1** [Telefonica/mistica-design#1373] [Proposals]
### Goal

To evaluate the proposal for a new file upload component to be added to the Mistica design system.

### Context

Currently the Mistica design system lacks a component, which allows the user to upload files, when needed for some actions like opening a ticket and attaching a file to it. The different teams would be developing their own solutions and this will lead to inconsistencies in the user experience.
The need in our project: For a new application, that is being used from internal employees, but also external partners and dealers, an upload function is needed, so the users can upload files, when they create tickets. For example: ID, contract pdf etc. 
Functions, that are needed (the functions are very specific for the case in our project, maybe they will be more generic for the component in Mistica):

- Upload button/component: The user can click on an icon to upload a file
- (Standard native features) For the upload the user can choose between taking a picture with the camera, choose a picture from the gallery or choose a file from his phone storage
- When selecting a file it should be checked for correct file extension(s) and possibly maximum file size (both customizable via props) 
- If the file is not valid (incorrect extension or file size too large) the component should show an error state 
- If the file is valid it should be uploaded to a customizable endpoint. The endpoint responds with JSON data (the server would save the file in a temporary store and return a temporary GUID)
- The JSON response from the endpoint should be stored in the form field state (when finally submitting the form, the temporary GUID can be used to permanently save the file on the server)
- After the user chooses the file/takes the picture, a preview is being shown, so the user knows which file he uploaded (preview image for images, file name for other types)
- On the preview there is an option to delete the chosen file

Here are some visuals to our proposal:
Upload button:
 
![Upload Button](https://github.com/Telefonica/mistica-design/assets/139142896/fa35ccfa-bc41-4a7a-8abb-9d5eb9ef28dd)

Preview of the uploaded file with a delete option:
 
![Upload preview](https://github.com/Telefonica/mistica-design/assets/139142896/0bb859c5-3606-4658-ac65-6d702856de62)

### Impact

Adding a standardized upload component would improve the user experience and reduce the development time for the teams, who must develop their own custom solutions for file uploading. It will also help new teams, who need the component in the future.

### Discussion Questions

•	Should the upload button link directly to the phone camera or storage, so the user can take a picture or choose a file or is there one more step needed, where the user should choose if he wants to take the picture or choose a file in the component? For example, a side bar is being opened with the options:
 
<img width="205" alt="Upload side bar" src="https://github.com/Telefonica/mistica-design/assets/139142896/0665b2c4-3f82-417e-8ed8-37116a631bd5">

•	Are there other use cases in other projects, where an upload component is needed?
•	Is a preview of the uploaded file preferred also for other projects or are there other options like name of the uploaded file?

### Additional considerations

The launch date of our projetct is in January 2024. Our goal will be to close the development in November 2023 and to use December to test and upload the app into the stores. So our team need to consider the timings of the development of the new upload component or find a different solution, if the component is not needed in the Mistica design system.

**Comment by HectorSancho** [comment]
Hi! we have explored several options because there were also requests from other teams.

We have solved it with rows boxed, I think it is not necessary to create a new component.

I think it is self explanatory and I have left a comment on the upload options, the easiest would be to go full native.

And sorry for the delay in replying. We have been very busy with work and with summer vacations in between.

Best regards.

https://www.figma.com/file/kColoW9UywPZYXmJ3axtK3/Upload-document?type=design&node-id=0%3A1&mode=design&t=rIV0oFD


![Captura de pantalla 2023-08-04 a las 13 32 36](https://github.com/Telefonica/mistica-design/assets/56122343/39855b64-0ca5-4bf8-a881-d7f30be14da4)


![Captura de pantalla 2023-08-04 a las 13 32 54](https://github.com/Telefonica/mistica-design/assets/56122343/0e399125-3f9c-44d7-9fe7-2771dc23f701)

**Reply by iveneva1** [reply]
Thank you a lot, Hector! This looks very good. I am sending it to our developers and I hope this solves our problem.

Best regards,
Ina

**Comment by yceballost** [comment]
Can I close this discussion @iveneva1 ?

**Reply by iveneva1** [reply]
Thank you! I closed it :)

**Comment by iveneva1** [comment]
We have a good solution. Thank you! I am closing the discussion
